### ISSUE TITLE
x/website/internal/dl: version sorting doesn't sort pre-release versions

### ISSUE THREAD

**Issue by nd** [golang/go#47367]
### What did you do?

Ran `curl "https://golang.org/dl/?mode=json&include=all"` in console

### What did you expect to see?

go1.17rc1 before go1.17beta1

### What did you see instead?

go1.17rc1 after go1.17beta1

**Comment by dmitshur** [comment]
Thanks for the report.

The `dl` package implements `fileOrder` sorting, and `versionLess` is used to compare two versions. When both are pre-release versions with same major, minor components, the pre-release part of the version isn't being sorted.

**Comment by zx2c4** [comment]
> When both are pre-release versions with same major, minor components, the pre-release part of the version isn't being sorted.

Hm, are you sure?

In `parseVersion`, the `tail` return value points to "beta1" or "rc1":

```go
func parseVersion(v string) (maj, min int, tail string) {
        if i := strings.Index(v, "beta"); i > 0 {
                tail = v[i:]
                v = v[:i]
        }
        if i := strings.Index(v, "rc"); i > 0 {
                tail = v[i:]
                v = v[:i]
        }
        p := strings.Split(strings.TrimPrefix(v, "go1."), ".")
        maj, _ = strconv.Atoi(p[0])
        if len(p) < 2 {
                return
        }
        min, _ = strconv.Atoi(p[1])
        return
}
```

Then in `versionLess`, the two tails are compared with `>=`:

```go
func versionLess(a, b string) bool {
        // Put stable releases first.
        if isStable(a) != isStable(b) {
                return isStable(a)
        }
        maja, mina, ta := parseVersion(a)
        majb, minb, tb := parseVersion(b)
        if maja == majb {
                if mina == minb {
                        return ta >= tb
                }
                return mina >= minb
        }
        return maja >= majb
}
```

This is then tested to be the case in `TestFileOrder`.

**Comment by zx2c4** [comment]
Ahh, actually what's happening is this, from `filesToReleases`:

```go
                        if len(unstable) != 0 {
                                // Only show one (latest) unstable version,
                                // consider the older ones to be archived.
                                archive = append(archive, *r)
                                return
                        }
```

Basically, since archives are currently first, and since there's only one unstable release, it's placed in the wrong spot for `include=all`, leading to the illusion that the sort was done wrong.

In other words, the problem of this issue is fixed by #50201 and https://golang.org/cl/371934.

**Comment by zx2c4** [comment]
I'll add a test case to that CL and make sure it works for the case here.
